File: java/src/main/java/implementation/creational/Builder.java
package implementation.creational;

import lombok.Getter;

/**
 * Builder encapsulates the construction of object and allows it to construct in steps
 */
public class Builder {
    void main() {
        Person person = Person.builder().setMale(false).setAge(1).build();
    }
}

class Person {
    private String name;
    private int age;
    private boolean isMale;

    public Person(PersonBuilder builder) {
        this.name = builder.getName();
        this.age = builder.getAge();
        this.isMale = builder.isMale();
    }

    @Getter
    static class PersonBuilder {

        private String name;
        private int age;
        private boolean isMale;

        public PersonBuilder setName(String name) {
            this.name = name;
            return this;
        }

        public PersonBuilder setAge(int age) {
            this.age = age;
            return this;
        }

        public PersonBuilder setMale(boolean male) {
            isMale = male;
            return this;
        }
        public Person build() {
            return new Person(this);
        }
    }

    static public PersonBuilder builder() {
        return new PersonBuilder();
    }
}

File: rust/src/creational/builder.rs
#[cfg(test)]
mod builder {
    pub struct Person {
        name: String,
        age: u8,
    }
    pub struct PersonBuilder {
        pub name: String,
        pub age: u8,
    }

    impl From<PersonBuilder> for Person {
        fn from(builder: PersonBuilder) -> Self {
            Person {
                name: builder.name,
                age: builder.age,
            }
        }
    }

    impl PersonBuilder {
        pub fn new() -> Self {
            PersonBuilder {
                name: String::new(),
                age: 0,
            }
        }

        pub fn set_name(mut self, name: &str) -> Self {
            self.name = name.to_string();
            self
        }

        pub fn set_age(mut self, age: u8) -> Self {
            self.age = age;
            self
        }

        pub fn build(self) -> Person {
            Person::from(self)
        }
    }

    impl Default for Person {
        fn default() -> Self {
            Person::from(PersonBuilder::default())
        }
    }

    impl Default for PersonBuilder {
        fn default() -> Self {
            PersonBuilder {
                name: String::from("John"),
                age: 18,
            }
        }
    }
    #[test]
    fn test_builder() {
        let person = PersonBuilder::new().set_name("John").set_age(18).build();
        assert_eq!(person.name, "John");
        assert_eq!(person.age, 18);
    }

    #[test]
    fn test_default() {
        let person = Person::default();
        assert_eq!(person.name, "John");
        assert_eq!(person.age, 18);
    }
}
